internal/agent/device/config/hook.go

```diff
@@ -0,0 +1,474 @@
+		}
+	}
+
+	for _, op := range action.Unit.Operations {
+		if err := executeSystemdOperation(ctx, systemdClient, op, actionTimeout, unitName); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func executeSystemdOperation(ctx context.Context, systemdClient *client.Systemd, op v1alpha1.ConfigHookActionSystemdUnitOperations, timeout time.Duration, unitName string) error {
+	ctx, cancel := context.WithTimeout(ctx, timeout)
+	defer cancel()
+
+	switch op {
+	case v1alpha1.SystemdStart:
+		if err := systemdClient.Start(ctx, unitName); err != nil {
+			return err
+		}
+	case v1alpha1.SystemdStop:
+		if err := systemdClient.Stop(ctx, unitName); err != nil {
+			return err
+		}
+	case v1alpha1.SystemdRestart:
+		if err := systemdClient.Restart(ctx, unitName); err != nil {
+			return err
+		}
+	case v1alpha1.SystemdReload:
+		if err := systemdClient.Reload(ctx, unitName); err != nil {
+			return err
+		}
+	case v1alpha1.SystemdEnable:
+		if err := systemdClient.Enable(ctx, unitName); err != nil {
+			return err
+		}
+	case v1alpha1.SystemdDisable:
+		if err := systemdClient.Disable(ctx, unitName); err != nil {
+			return err
+		}
+	case v1alpha1.SystemdDaemonReload:
+		if err := systemdClient.DaemonReload(ctx); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func handleHookActionExecutable(ctx context.Context, exec executer.Executer, action *v1alpha1.ConfigHookActionExecutableSpec, configFilePath string) error {
+	actionTimeout, err := parseTimeout(action.Timeout)
+	if err != nil {
+		return err
+	}
+
+	ctx, cancel := context.WithTimeout(ctx, actionTimeout)
+	defer cancel()
+
+	dirExists, err := dirExists(action.Executable.WorkDir)
+	if err != nil {
+		return err
+	}
+
+	// we expect the directory to exist should be created by config if its new.
+	if !dirExists {
+		return os.ErrNotExist
+	}
+
+	// replace file token in args if it exists
+	// TODO: cache the map.
+	tokenMap := newTokenMap(configFilePath)
+	args, err := replaceTokensInArgs(action.Executable.Args, tokenMap)
+	if err != nil {
+		return err
+	}
+
+	var envVars []string
+	if action.Executable.EnvVars != nil {
+		envVars = *action.Executable.EnvVars
+	}
+
+	_, stderr, exitCode := exec.ExecuteWithContextFromDir(ctx, action.Executable.WorkDir, action.Executable.Command, args, envVars...)
+	if exitCode != 0 {
+		return fmt.Errorf("failed to execute command: %s %d: %s", action.Executable.Command, exitCode, stderr)
+	}
+
+	return nil
+}
+
+// addOrReplaceHookHandler adds or replaces a hook handler in the manager. this function assumes a lock is held.
+func addOrReplaceHookHandler(fsMonitor FileMonitor, newHook *v1alpha1.DeviceConfigHookSpec, existingHandlers map[string]*HookHandler) error {
+	// build lookup for file operations
+	opActions := make(map[fsnotify.Op][]v1alpha1.ConfigHookAction)
+	for _, action := range newHook.Actions {
+		hookActionType, err := action.Discriminator()
+		if err != nil {
+			return err
+		}
+		switch hookActionType {
+		case SystemdHookActionType:
```

Comment:
The code for `SystemdHookActionType` and  `ExecutableHookActionType` looks exactly the same.  Can they be unified to a single `case` section ?